cmds/exp/cpu/doc.go

```diff
@@ -126,4 +148,74 @@
 //    -rw-r--r-- 1 rminnich rminnich 577 May 17 17:06 /home/rminnich/xyz
 //    rminnich@xcpu:~/gopath/src/github.com/u-root/u-root/xcmds/cpu$
+//
+// A note on sizes of things. We can get an image down to 3 MiB if the only
+// binary is cpu. See github.com:linuxboot/mainboards/aeeon/up for an example.
+//
+// You may see that the remote cpu writes a nonce back to the client, and wonder why
+// that is. This gets back to how we make the client name space available to remote processes.
+// To export the client name space to the remote cpu proc
+// we use a remote ssh forward so the remote kernel can mount
+// our built-in server over 9p. But how do we ensure the mount comes
+// from the remote cpu proc and not something else?
+// We need to ensure that the socket has only
+// been opened by the remote cpu process.
+// The way we do it here is the result of discussions with Eric Grosse, among others.
+// Not that he endorses this solution, but he raised awareness
+// of the issues.
+//
+// The 'remote' invocation of CPU is designed to work under
+// cpu acting as init OR an sshd (i.e. ssh host cpu -r ...)
+// Anything that makes the cpu init mode special is out; we would also have
+// to change sshd to do the same things and we can't change
+// every ssh server in the world. For anoher example, using Linux
+// private network name spaces for the ssh forward is out,
+// as not all ssh daemons support that.
+// Trying to use ssh-forwarded Unix domain sockets is out, as we would want them placed in the
+// private name space of the remote cpu process, and no ssh daemons support that; further
+// the Go ssh packages don't support Unix domain sockets. Even were we to add them (I did)
+// the approach fails as we can't specify that they be in a private name space.
+// We'd rather not do a full TLS transaction on the forwarded port,
+// since we for sure don't want another layer of encryption running
+// over our already-encrypted, already-authenticated ssh connection.
+//
+// In the last year we've experimented with several approaches that all failed
+// due to limits in how ssh protocol works, how the ssh packages work, how various ssh daemons work,
+// or how Linux manages name spaces. What we describe below is the simplest approach
+// that works on everything we've tried.
+//
+// We only do one accept on the client side.
+// For the kernel 9p mount at the remote, we use the fd transport.
+// The fd we pass is for a socket that has been verified.
+// We verify as follows:
+// client generates a nonce and adds it as an environment variable (not argv!).
+// We don't put the nonce in argv as an adversary could then see it in ps.
+// The remote cpu process reads that variable and removes it from the environment.
+// The remote cpu process writes that nonce back on the port forward socket within 10 ms.
+// The 10ms requirement is to defend (imperfectly) against a bad person running the remote
```

Comment:
I don't understand how we can trust the remote at all if a bad person can run ptrace on it or why a 10ms deadline helps.

Reply:
Once the localhost tcp socket appears there is a race: someone can read my environment and write the nonce instead of me. This is because we have no good way to set up a private IPC channel, usable by the kernel for a 9p mount, via the ssh packages we have today. Were we able to set up such a channel we would not even need the nonce: we could pipe() and use the fds for the mount.

This approach is far from perfect. The deadline is sort-of-ok proof of someone running at my privilege against someone jumping in once they see the socket appear and before I can write the nonce back. I've talked to someone who does this for a living and they feel it's a reasonable approach. We can talk more later. 

What I wanted was a way to set up an ssh channel, which I could connect to a pipe and use for the 9p mount. I could not find a way to do it. If you can think of one let me know. All I need is a way to set up that client, much as stderr is set up now, and all this nonsense goes away.